crates/bevy_reflect/src/utility.rs

```diff
@@ -0,0 +1,148 @@
+/// use bevy_reflect::utility::TypeInfoCell;
+///
+/// struct Foo;
+///
+/// impl Typed for Foo {
+///   fn type_info() -> &'static TypeInfo {
+///     static CELL: TypeInfoCell = TypeInfoCell::non_generic();
+///     CELL.get_or_insert::<Self, _>(|| {
+///       let info = ValueInfo::new::<Self>();
+///       TypeInfo::Value(info)
+///     })
+///   }
+/// }
+/// #
+/// # unsafe impl Reflect for Foo {
+/// #   fn type_name(&self) -> &str { todo!() }
+/// #   fn get_type_info(&self) -> &'static TypeInfo { todo!() }
+/// #   fn any(&self) -> &dyn Any { todo!() }
+/// #   fn any_mut(&mut self) -> &mut dyn Any { todo!() }
+/// #   fn as_reflect(&self) -> &dyn Reflect { todo!() }
+/// #   fn as_reflect_mut(&mut self) -> &mut dyn Reflect { todo!() }
+/// #   fn apply(&mut self, value: &dyn Reflect) { todo!() }
+/// #   fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> { todo!() }
+/// #   fn reflect_ref(&self) -> ReflectRef { todo!() }
+/// #   fn reflect_mut(&mut self) -> ReflectMut { todo!() }
+/// #   fn clone_value(&self) -> Box<dyn Reflect> { todo!() }
+/// # }
+/// ```
+///
+/// ## Generic
+///
+/// For generic types, [`TypeInfoCell`] should be initialized via the [`generic`]
+/// method. This will store multiple instances of [`TypeInfo`], accessible by [`TypeId`].
+///
+/// This allows `Foo<T>` to use the same [`TypeInfoCell`] for monomorphized type.
+///
+/// ```
+/// # use std::any::Any;
+/// # use std::marker::PhantomData;
+/// # use bevy_reflect::{Reflect, ReflectMut, ReflectRef, Typed, TypeInfo, ValueInfo};
+/// use bevy_reflect::utility::TypeInfoCell;
+///
+/// struct Foo<T: Reflect>(PhantomData<T>);
+///
+/// impl<T: Reflect> Typed for Foo<T> {
+///   fn type_info() -> &'static TypeInfo {
+///     static CELL: TypeInfoCell = TypeInfoCell::generic();
+///     CELL.get_or_insert::<Self, _>(|| {
+///       let info = ValueInfo::new::<Self>();
+///       TypeInfo::Value(info)
+///     })
+///   }
+/// }
+/// #
+/// # unsafe impl<T: Reflect> Reflect for Foo<T> {
+/// #   fn type_name(&self) -> &str { todo!() }
+/// #   fn get_type_info(&self) -> &'static TypeInfo { todo!() }
+/// #   fn any(&self) -> &dyn Any { todo!() }
+/// #   fn any_mut(&mut self) -> &mut dyn Any { todo!() }
+/// #   fn as_reflect(&self) -> &dyn Reflect { todo!() }
+/// #   fn as_reflect_mut(&mut self) -> &mut dyn Reflect { todo!() }
+/// #   fn apply(&mut self, value: &dyn Reflect) { todo!() }
+/// #   fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> { todo!() }
+/// #   fn reflect_ref(&self) -> ReflectRef { todo!() }
+/// #   fn reflect_mut(&mut self) -> ReflectMut { todo!() }
+/// #   fn clone_value(&self) -> Box<dyn Reflect> { todo!() }
+/// # }
+/// ```
+///
+/// [`once_cell`]: https://docs.rs/once_cell/latest/once_cell/
+/// [`non_generic`]: TypeInfoCell::non_generic
+/// [`generic`]: TypeInfoCell::generic
+pub struct TypeInfoCell(TypeInfoCellType);
```

Comment:
Thoughts on the name `TypeInfoCell`? I wasn't sure if this was "correct" or not.